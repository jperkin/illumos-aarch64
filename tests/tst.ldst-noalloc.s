/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.3.7 Load/store no-allocate pair (offset)
 *
 * C5.6.176 STNP
 * C5.6.80  LDNP
 * C6.3.283 STNP (SIMD&FP)
 * C6.3.164 LDNP (SIMD&FP)
 *
 * Test the minimum/maximum of registers and signed imm7.
 */
.macro ldstnoalloc1 op, reg, size, simd=0
	\op	\reg\()0,  \reg\()0,  [x0]
	\op	\reg\()30, \reg\()30, [x30, #(\size*63)]
	.if \simd
	\op	\reg\()31, \reg\()31, [sp,  #(\size*-64)]
	.else
	\op	\reg\()zr, \reg\()zr, [sp,  #(\size*-64)]
	.endif
.endm

.macro ldstnoalloc op
	ldstnoalloc1 \op, w,  4,
	ldstnoalloc1 \op, x,  8,
	ldstnoalloc1 \op, s,  4, 1
	ldstnoalloc1 \op, d,  8, 1
	ldstnoalloc1 \op, q, 16, 1
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	ldstnoalloc stnp
	ldstnoalloc ldnp
.size libdis_test, [.-libdis_test]
