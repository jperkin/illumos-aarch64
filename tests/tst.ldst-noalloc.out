    libdis_test:      00 00 00 28  stnp w0, w0, [x0]
    libdis_test+0x4:  de fb 1f 28  stnp w30, w30, [x30, #252]
    libdis_test+0x8:  ff 7f 20 28  stnp wzr, wzr, [sp, #-256]
    libdis_test+0xc:  00 00 00 a8  stnp x0, x0, [x0]
    libdis_test+0x10: de fb 1f a8  stnp x30, x30, [x30, #504]
    libdis_test+0x14: ff 7f 20 a8  stnp xzr, xzr, [sp, #-512]
    libdis_test+0x18: 00 00 00 2c  stnp s0, s0, [x0]
    libdis_test+0x1c: de fb 1f 2c  stnp s30, s30, [x30, #252]
    libdis_test+0x20: ff 7f 20 2c  stnp s31, s31, [sp, #-256]
    libdis_test+0x24: 00 00 00 6c  stnp d0, d0, [x0]
    libdis_test+0x28: de fb 1f 6c  stnp d30, d30, [x30, #504]
    libdis_test+0x2c: ff 7f 20 6c  stnp d31, d31, [sp, #-512]
    libdis_test+0x30: 00 00 00 ac  stnp q0, q0, [x0]
    libdis_test+0x34: de fb 1f ac  stnp q30, q30, [x30, #1008]
    libdis_test+0x38: ff 7f 20 ac  stnp q31, q31, [sp, #-1024]
    libdis_test+0x3c: 00 00 40 28  ldnp w0, w0, [x0]
    libdis_test+0x40: de fb 5f 28  ldnp w30, w30, [x30, #252]
    libdis_test+0x44: ff 7f 60 28  ldnp wzr, wzr, [sp, #-256]
    libdis_test+0x48: 00 00 40 a8  ldnp x0, x0, [x0]
    libdis_test+0x4c: de fb 5f a8  ldnp x30, x30, [x30, #504]
    libdis_test+0x50: ff 7f 60 a8  ldnp xzr, xzr, [sp, #-512]
    libdis_test+0x54: 00 00 40 2c  ldnp s0, s0, [x0]
    libdis_test+0x58: de fb 5f 2c  ldnp s30, s30, [x30, #252]
    libdis_test+0x5c: ff 7f 60 2c  ldnp s31, s31, [sp, #-256]
    libdis_test+0x60: 00 00 40 6c  ldnp d0, d0, [x0]
    libdis_test+0x64: de fb 5f 6c  ldnp d30, d30, [x30, #504]
    libdis_test+0x68: ff 7f 60 6c  ldnp d31, d31, [sp, #-512]
    libdis_test+0x6c: 00 00 40 ac  ldnp q0, q0, [x0]
    libdis_test+0x70: de fb 5f ac  ldnp q30, q30, [x30, #1008]
    libdis_test+0x74: ff 7f 60 ac  ldnp q31, q31, [sp, #-1024]
